### DISCUSSION TITLE
MBP M1 fails to 'brew install' but 'brew update' still work.

### DISCUSSION THREAD

**Discussion by victranit** [Homebrew/discussions#716] [Everyday usage] [answered]
**I have an issues with `homebrew` in my MBP 2020 M1 as below.**

When i run :
`brew update` --> OK
`brew install XXX` --> ERROR
`brew upgrade` --> ERROR
`brew doctor` --> ERROR

```
taitq@Tais-MacBook-Pro-M1 apache % brew update
Already up-to-date.
taitq@Tais-MacBook-Pro-M1 apache % brew 
```

```
taitq@Tais-MacBook-Pro-M1 apache % brew install mysql
Traceback (most recent call last):
	22: from /opt/homebrew/Library/Homebrew/brew.rb:31:in `<main>'
	21: from /opt/homebrew/Library/Homebrew/brew.rb:31:in `require_relative'
	20: from /opt/homebrew/Library/Homebrew/global.rb:23:in `<top (required)>'
	19: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
	18: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
	17: from /opt/homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/activesupport-6.1.1/lib/active_support/core_ext/numeric/time.rb:3:in `<top (required)>'
	16: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
	15: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
	14: from /opt/homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/activesupport-6.1.1/lib/active_support/duration.rb:3:in `<top (required)>'
	13: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
	12: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
	11: from /opt/homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/activesupport-6.1.1/lib/active_support/core_ext/array/conversions.rb:3:in `<top (required)>'
	10: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
	 9: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
	 8: from /opt/homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/activesupport-6.1.1/lib/active_support/xml_mini.rb:3:in `<top (required)>'
	 7: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
	 6: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
	 5: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/time.rb:3:in `<top (required)>'
	 4: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
	 3: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
	 2: from /Library/Ruby/Gems/2.6.0/gems/date-3.1.1/lib/date.rb:4:in `<top (required)>'
	 1: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require': dlopen(/Library/Ruby/Gems/2.6.0/gems/date-3.1.1/lib/date_core.bundle, 0x0009): missing compatible arch in /Library/Ruby/Gems/2.6.0/gems/date-3.1.1/lib/date_core.bundle - /Library/Ruby/Gems/2.6.0/gems/date-3.1.1/lib/date_core.bundle (LoadError)
```

AND I just uninstalled my gem date 3.1.1

```
taitq@Tais-MacBook-Pro-M1 apache % gem uninstall date 
Gem date-3.0.0 cannot be uninstalled because it is a default gem
taitq@Tais-MacBook-Pro-M1 apache % 
```

```
taitq@Tais-MacBook-Pro-M1 apache % gem list

*** LOCAL GEMS ***

actioncable (6.1.1, 6.0.3.4)
actionmailbox (6.1.1, 6.0.3.4)
actionmailer (6.1.1, 6.0.3.4)
actionpack (6.1.1, 6.0.3.4)
actiontext (6.1.1, 6.0.3.4)
actionview (6.1.1, 6.0.3.4)
active_model_serializers (0.10.12)
activejob (6.1.1, 6.0.3.4)
activemodel (6.1.1, 6.0.3.4)
activerecord (6.1.1, 6.0.3.4)
activestorage (6.1.1, 6.0.3.4)
activesupport (6.1.1, 6.0.3.4, 5.2.4.4)
addressable (2.7.0)
algoliasearch (1.27.5)
ast (2.4.2, 2.4.1)
atomos (0.1.3)
bcrypt (3.1.16)
benchmark (0.1.1, default: 0.1.0)
bigdecimal (3.0.0, default: 2.0.0)
bindex (0.8.1)
bootsnap (1.7.1, 1.7.0, 1.5.1)
builder (3.2.4)
bundler (default: 2.2.8)
byebug (11.1.3)
capybara (3.35.3, 3.34.0)
case_transform (0.2)
CFPropertyList (3.0.3)
cgi (0.2.0, default: 0.1.0)
childprocess (4.0.0, 3.0.0)
claide (1.0.3)
cocoapods (1.10.1)
cocoapods-clean (0.0.1)
cocoapods-core (1.10.1)
cocoapods-deintegrate (1.0.4)
cocoapods-downloader (1.4.0)
cocoapods-plugins (1.0.0)
cocoapods-search (1.0.0)
cocoapods-trunk (1.5.0)
cocoapods-try (1.2.0)
coderay (1.1.3)
colored2 (3.1.2)
concurrent-ruby (1.1.8, 1.1.7)
config (2.2.3)
connection_pool (2.2.3)
crass (1.0.6)
csv (3.1.9, default: 3.1.2)
database_cleaner (2.0.1, 2.0.0, 1.8.5)
database_cleaner-active_record (2.0.0)
database_cleaner-core (2.0.1, 2.0.0)
date (default: 3.0.0)
dbm (default: 1.1.0)
deep_merge (1.2.1)
delegate (0.2.0, default: 0.1.0)
did_you_mean (1.5.0, default: 1.4.0)
diff-lcs (1.4.4)
digest (3.0.0)
dry-configurable (0.12.0)
dry-container (0.7.2)
dry-core (0.5.0)
dry-equalizer (0.3.0)
dry-inflector (0.2.0)
dry-initializer (3.0.4)
dry-logic (1.1.0)
dry-schema (1.6.1, 1.5.6)
dry-types (1.5.0, 1.4.0)
dry-validation (1.6.0)
erubi (1.10.0)
escape (0.0.4)
etc (1.2.0, default: 1.1.0)
ethon (0.12.0)
factory_bot (6.1.0)
factory_bot_rails (6.1.0)
faker (2.15.1)
fcntl (default: 1.0.0)
ffi (1.14.2)
fiddle (1.0.7, default: 1.0.0)
figaro (1.2.0)
fileutils (1.5.0, default: 1.4.1)
forwardable (1.3.2, default: 1.3.1)
fourflusher (2.3.1)
fuzzy_match (2.1.0, 2.0.4)
getoptlong (0.1.1, default: 0.1.0)
gh_inspector (1.1.3)
globalid (0.4.2)
httpclient (2.8.3)
i18n (1.8.8, 1.8.7)
io-console (0.5.7, default: 0.5.6)
ipaddr (default: 1.2.2)
irb (1.3.2, default: 1.2.6)
jbuilder (2.11.2, 2.10.1)
json (2.5.1, default: 2.3.0)
jsonapi-renderer (0.2.2)
kaminari (1.2.1)
kaminari-actionview (1.2.1)
kaminari-activerecord (1.2.1)
kaminari-core (1.2.1)
listen (3.4.1)
logger (1.4.3, default: 1.4.2)
loofah (2.9.0)
mail (2.7.1)
marcel (0.3.3)
matrix (0.4.1, default: 0.2.0)
method_source (1.0.0)
mimemagic (0.3.5)
mini_magick (4.11.0)
mini_mime (1.0.2)
mini_portile2 (2.5.0)
minitest (5.14.3, 5.13.0)
molinillo (0.7.0, 0.6.6)
msgpack (1.4.2, 1.3.3)
mutex_m (0.1.1, default: 0.1.0)
mysql2 (0.5.3)
nanaimo (0.3.0)
nap (1.1.0)
net-pop (0.1.1, default: 0.1.0)
net-protocol (0.1.0)
net-smtp (0.2.1, default: 0.1.0)
net-telnet (0.2.0)
netrc (0.11.0)
nio4r (2.5.4)
nokogiri (1.11.1 arm64-darwin)
observer (0.1.1, default: 0.1.0)
open3 (0.1.1, default: 0.1.0)
openssl (2.2.0, default: 2.1.2)
ostruct (0.3.3, default: 0.2.0)
pagy (3.10.0)
parallel (1.20.1)
paranoia (2.4.3)
parser (3.0.0.0)
power_assert (2.0.0, 1.1.7)
prime (0.1.2, default: 0.1.1)
pry (0.13.1)
pry-byebug (3.9.0)
pstore (0.1.1, default: 0.1.0)
psych (3.3.0, default: 3.1.0)
public_suffix (4.0.6)
puma (5.2.0, 4.3.7)
racc (1.5.2, default: 1.4.16)
rack (2.2.3)
rack-attack (6.4.0)
rack-cors (1.1.1)
rack-test (1.1.0)
rails (6.1.1, 6.0.3.4)
rails-dom-testing (2.0.3)
rails-html-sanitizer (1.3.0)
railties (6.1.1, 6.0.3.4)
rainbow (3.0.0)
rake (13.0.3, 13.0.1)
ransack (2.4.2)
rb-fsevent (0.10.4)
rb-inotify (0.10.1)
rdoc (6.3.0, default: 6.2.1)
readline (default: 0.0.2)
readline-ext (0.1.1, default: 0.1.0)
redis (4.2.5)
regexp_parser (2.0.3, 1.8.2)
reline (0.2.2, default: 0.1.5)
rexml (3.2.4, default: 3.2.3)
rspec-core (3.10.1)
rspec-expectations (3.10.1)
rspec-mocks (3.10.2)
rspec-rails (4.0.2)
rspec-support (3.10.2)
rss (0.2.9, default: 0.2.8)
rubocop (1.9.1)
rubocop-ast (1.4.1)
rubocop-rails (2.9.1)
ruby-macho (2.5.0, 1.4.0)
ruby-progressbar (1.11.0)
rubygems-update (3.2.8)
rubyzip (2.3.0)
sdbm (default: 1.0.0)
selenium-webdriver (3.142.7)
shoulda-matchers (4.5.1)
sidekiq (6.1.3)
singleton (0.1.1, default: 0.1.0)
spring (2.1.1)
spring-watcher-listen (2.0.1)
sprockets (4.0.2)
sprockets-rails (3.2.2)
stringio (3.0.0, default: 0.1.0)
strscan (3.0.0, default: 1.0.3)
swagger-blocks (3.0.0)
test-unit (3.4.0, 3.3.4)
thor (1.1.0, 1.0.1)
thread_safe (0.3.6)
timecop (0.9.2)
timeout (0.1.1, default: 0.1.0)
tracer (0.1.1, default: 0.1.0)
typhoeus (1.4.0)
tzinfo (2.0.4, 1.2.9)
unicode-display_width (2.0.0)
uri (0.10.1, default: 0.10.0)
web-console (4.1.0)
webdrivers (4.5.0)
webrick (1.7.0, default: 1.6.0)
websocket-driver (0.7.3)
websocket-extensions (0.1.5)
xcodeproj (1.19.0)
xmlrpc (0.3.1, 0.3.0)
xpath (3.2.0)
yaml (0.1.1, default: 0.1.0)
zeitwerk (2.4.2)
zlib (default: 1.1.0)
```

But when i rerun `brew install XXX` again, i still get the same error! (about gem date 3.1.1)
I have no ideas, can anyone please help me?
Thanks!

**Comment by Bo98** [comment, marked as answer]
Try the same but with `/usr/bin/gem` explicitly, just in case it's picking up another Ruby installation.

**Reply by victranit** [reply]
You are my Hero!!! 🥇 

I uninstalled these gems : **date**, **etc**, **io-console**, **bigdecimal** by `/usr/bin/gem`.
And now my homebrew work fine!!!
Thanks so much!

**Comment by trx222** [comment]
I also tried but now ruby/gem seems to be broken - I can not reinstall homebrew or system ruby.

If I try to reinstall I always get:

/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require': dlopen(/Library/Ruby/Gems/2.6.0/gems/psych-3.3.0/lib/psych.bundle, 0x0009): missing compatible arch in /Library/Ruby/Gems/2.6.0/gems/psych-3.3.0/lib/psych.bundle - /Library/Ruby/Gems/2.6.0/gems/psych-3.3.0/lib/psych.bundle (LoadError)

If I switch ruby with rbenv to an other version gem and ruby works fine but the homebrew installer still uses the system version and I again get this message. I stuck with it over days now and dont find a solution :-(

**Reply by Bo98** [reply]
What happens when you do `/usr/bin/gem uninstall psych`?

**Reply by trx222** [reply]
Same error:

`
% /usr/bin/gem uninstall psych

/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/yaml.rb:3: warning: It seems your ruby installation is missing psych (for YAML output).
To eliminate this warning, please install libyaml and reinstall your ruby.
Traceback (most recent call last):
	14: from /usr/bin/gem:21:in `<main>'
	13: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/gem_runner.rb:44:in `run'
	12: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/gem_runner.rb:79:in `do_configuration'
	11: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/gem_runner.rb:79:in `new'
	10: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/config_file.rb:198:in `initialize'
	 9: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/config_file.rb:346:in `load_file'
	 8: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems.rb:701:in `load_yaml'
	 7: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
	 6: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
	 5: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/yaml.rb:4:in `<top (required)>'
	 4: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
	 3: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
	 2: from /Library/Ruby/Gems/2.6.0/gems/psych-3.3.0/lib/psych.rb:13:in `<top (required)>'
	 1: from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require': dlopen(/Library/Ruby/Gems/2.6.0/gems/psych-3.3.0/lib/psych.bundle, 0x0009): missing compatible arch in /Library/Ruby/Gems/2.6.0/gems/psych-3.3.0/lib/psych.bundle - /Library/Ruby/Gems/2.6.0/gems/psych-3.3.0/lib/psych.bundle (LoadError)
`

**Reply by Bo98** [reply]
Ah a little chicken-and-egg problem because `gem` loads `psych`.

Maybe try:
```
rm /Library/Ruby/Gems/2.6.0/gems/psych-3.3.0/lib/psych.bundle
/usr/bin/gem uninstall psych
```

**Reply by trx222** [reply]
That was the key! Thank you very much!

I had to do:
> rm /Library/Ruby/Gems/2.6.0/gems/psych-3.3.0/lib/psych.bundle
> rm /Library/Ruby/Gems/2.6.0/extensions/universal-darwin-20/2.6.0/psych-3.3.0/psych.bundle
> gem uninstall psych

and then the same with

> rm /Library/Ruby/Gems/2.6.0/gems/strscan-3.0.0/lib/strscan.bundle
> rm /Library/Ruby/Gems/2.6.0/extensions/universal-darwin-20/2.6.0/strscan-3.0.0/strscan.bundle
> gem uninstall strscan

after this, gem worked again, homebrew installation successfull. I still hat to unsinstall date, io-console, etc and zlib but the I was able to install brew apps like mc

:) 👍  

> Ah a little chicken-and-egg problem because gem loads psych.

exactly ;)

**Reply by trx222** [reply]
Some hints ... I got back the problem with 
> gem update

So be carefull with this ..

**Reply by Bo98** [reply]
Hmm, interesting that system Ruby is installing gems that aren't properly compiled for arm64.

Unfortunately, it's more Apple's bug than Homebrew's.

**Reply by trx222** [reply]
Thats true. It just comes up when you need homebrew. I will try to work with different ruby versions now, to not run again into gem update problem.

**Comment by CrushDope** [comment]
I use MBP first , how can i install homebrew in M1 mbp,Thanks

**Reply by carlocab** [reply]
Have you tried following the instructions at https://brew.sh?
